File: userservice/src/main/java/hu/bme/aut/classifiedadvertisementsite/userservice/security/AuthHeaderFilter.java
package hu.bme.aut.classifiedadvertisementsite.userservice.security;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;

@Slf4j
public class AuthHeaderFilter extends OncePerRequestFilter {

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        String authHeader = request.getHeader("x-user-data");

        if (authHeader != null) {
            ObjectMapper mapper = new ObjectMapper();
            JsonNode node;
            try {
                node = mapper.readTree(Base64.getDecoder().decode(authHeader));

                UserDetails userDetails = new UserDetailsImpl(
                        node.get("id").asInt(),
                        node.get("username").asText(),
                        node.get("email").asText(),
                        null,
                        Boolean.TRUE,
                        jsonArrayToList(node.get("roles")));

                UsernamePasswordAuthenticationToken authentication = new UsernamePasswordAuthenticationToken(userDetails, null, userDetails.getAuthorities());

                authentication.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));

                SecurityContextHolder.getContext().setAuthentication(authentication);
            } catch (Exception e) {
                log.error("Cannot set user authentication: {}", e.getMessage());
            }
        }

        filterChain.doFilter(request, response);
    }

    private List<String> jsonArrayToList(JsonNode node) {
        List<String> list = new ArrayList<>();
        if (node.isArray()) {
            for (final JsonNode objNode : node) {
                list.add(objNode.asText());
            }
        }
        return list;
    }
}

File: advertisementservice/src/main/kotlin/hu/bme/aut/classifiedadvertisementsite/advertisementservice/security/AuthHeaderFilter.kt
package hu.bme.aut.classifiedadvertisementsite.advertisementservice.security

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource
import org.springframework.web.filter.OncePerRequestFilter
import java.util.*
import kotlin.collections.ArrayList

class AuthHeaderFilter : OncePerRequestFilter() {

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        val authHeader = request.getHeader("x-user-data")

        if (authHeader != null) {
            val mapper = ObjectMapper()
            val node: JsonNode
            try {
                node = mapper.readTree(Base64.getDecoder().decode(authHeader))
                val userDetails: UserDetails = UserDetailsImpl(
                    node["id"].asInt(),
                    node["username"].asText(),
                    node["email"].asText(),
                    true,
                    jsonArrayToList(node["roles"])
                )
                val authentication =
                    UsernamePasswordAuthenticationToken(userDetails, null, userDetails.authorities)
                authentication.details = WebAuthenticationDetailsSource().buildDetails(request)
                SecurityContextHolder.getContext().authentication = authentication
            } catch (e: Exception) {
                TODO("logging")
            }
        }

        filterChain.doFilter(request, response)
    }

    private fun jsonArrayToList(node: JsonNode): List<String> {
        val list: MutableList<String> = ArrayList()
        if (node.isArray) {
            for (objNode in node) {
                list.add(objNode.asText())
            }
        }
        return list
    }
}